smart-contracts/osmosis/contracts/cl-vault/src/vault/swap.rs

```diff
@@ -30,51 +28,82 @@ pub fn execute_swap_non_vault_funds(
     let mut swap_msgs: Vec<CosmosMsg> = vec![];
 
     for swap_operation in swap_operations {
-        let token_in_denom = swap_operation.token_in_denom.clone();
-        let pool_token_0 = pool_config.token0.clone();
-        let pool_token_1 = pool_config.token1.clone();
+        let token_in_denom = &swap_operation.token_in_denom;
 
-        if token_in_denom == pool_token_0 || token_in_denom == pool_token_1 {
+        if token_in_denom == &pool_config.token0 || token_in_denom == &pool_config.token1 {
             return Err(ContractError::InvalidSwapAssets {});
         }
 
         let balance_in_contract = deps
             .querier
-            .query_balance(
-                contract_address.clone(),
-                swap_operation.clone().token_in_denom,
-            )?
+            .query_balance(env.clone().contract.address, token_in_denom.clone())?
             .amount;
-
-        if balance_in_contract == Uint128::zero() {
-            return Err(ContractError::InsufficientFunds {});
+        if balance_in_contract.is_zero() {
+            return Err(ContractError::InsufficientFundsForSwap {
+                balance: balance_in_contract,
+                needed: Uint128::new(1),
+            });
         }
 
-        let part_0_amount = balance_in_contract.checked_div(Uint128::new(2))?;
-        let part_1_amount = balance_in_contract
-            .checked_add(Uint128::new(1))?
-            .checked_div(Uint128::new(2))?;
-
-        let token_out_min_amount_0 =
-            part_0_amount.checked_mul_floor(vault_config.swap_max_slippage)?;
-        let token_out_min_amount_1 =
-            part_1_amount.checked_mul_floor(vault_config.swap_max_slippage)?;
+        // Get the current position balance ratio to compute the amount of external funds we want to swap into either token0 or token1 from the vault's pool
+        let position_balance = get_position_balance(deps.storage, &deps.querier)?;
+        let to_token0_amount = Uint128::from(
+            (balance_in_contract.u128() as f64
+                * position_balance.0.to_string().parse::<f64>().unwrap()) as u128,
+        );
+        let to_token1_amount = Uint128::from(
+            (balance_in_contract.u128() as f64
+                * position_balance.1.to_string().parse::<f64>().unwrap()) as u128,
+        );
 
         let dex_router = DEX_ROUTER.may_load(deps.storage)?;
+
+        // TODO: Validate that the swap_operation.pool_id_base is about token_in_denom and pool_config.token0 assets or throw error
```

Comment:
This is probably unnecessary as hijacking the pool_id wouldn't go through as we ensure that the min_receive token is: `coin(token_out_min_amount_base.into(), pool_config.clone().token0)`. So we would get an error like: `called `Result::unwrap()` on an `Err` value: ExecuteError { msg: "failed to execute message; message index: 0: dispatch: submessages: tokenIn (udenom) does not match any asset in pool" }` from Osmosis protocol. What do you think @lubkoll ?

Reply:
I wouldn't check this. In case you want to keep a TODO open, I'd prefer a ticket to a comment.